api/v1/coupons.js

```diff
@@ -30,6 +32,38 @@ var createCouponForNewUser = function (req, res, next) {
 };
 exports.createCouponForNewUser = createCouponForNewUser;
 
+var getDiscountOrderValueByCouponID = function (req, res, next) {
+  var couponId = req.params.couponID;
+  var username = req.query.username;
+  var orderValue = req.query.orderValue;
+
+  var isBelongToUsers = coupon.isBelongToUsers(couponId, username);
+  var isCouponValid = coupon.isCouponValid(couponId);
+  
+  Promise.join(isBelongToUsers, isCouponValid, function (belong, valid) {
+    if (!belong && valid) return true;
+    return false;
+  }).then(function (ableToUse) {
+    if (ableToUse) {
+        return coupon.getCouponCodesByCouponCode(couponId).then(function (coupons) {
+        return coupons[0];
+      }).then(function (couponObject) {
+        var ruleType = couponObject.couponRule.type;
+        var ruleValue = couponObject.couponRule.value;
+        console.log(ruleType + '---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------')
+        console.log(ruleValue + '---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------')
+        console.log(orderValue + '---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------')
+
+        var discountedValue = coupon.getDiscountedValue(ruleType, ruleValue, orderValue);
+        res.statusCode = 201;
```

Comment:
chained promises are not nested, they would look like this:
```
remotedb.allDocs(...).then(function (resultOfAllDocs) {
  return localdb.put(...);
}).then(function (resultOfPut) {
  return localdb.get(...);
}).then(function (resultOfGet) {
  return localdb.put(...);
}).catch(function (err) {
  console.log(err);
});

```
And this is the same code but a bad example

```
remotedb.allDocs({
  include_docs: true,
  attachments: true
}).then(function (result) {
  var docs = result.rows;
  docs.forEach(function(element) {
    localdb.put(element.doc).then(function(response) {
      alert("Pulled doc with id " + element.doc._id + " and added to local db.");
    }).catch(function (err) {
      if (err.name == 'conflict') {
        localdb.get(element.doc._id).then(function (resp) {
          localdb.remove(resp._id, resp._rev).then(function (resp) {
```

Have a read at this https://pouchdb.com/2015/05/18/we-have-a-problem-with-promises.html